packages/datadog-plugin-fs/src/index.js

```diff
@@ -0,0 +1,497 @@
+    const tags = makeFSTags(fd, null, null, config, tracer)
+    tags['file.mode'] = '0o' + mode.toString(8)
+    return tags
+  }
+}
+
+function createPathTags (config, tracer) {
+  return function pathTags (path) {
+    if (typeof path === 'number') {
+      return
+    }
+    return makeFSTags(path, null, null, config, tracer)
+  }
+}
+
+function createFDTags (config, tracer) {
+  return function fdTags (fd) {
+    if (typeof this === 'object' && this !== null && this.fd) {
+      fd = this.fd
+    }
+    if (typeof fd !== 'number') {
+      return
+    }
+    return makeFSTags(fd, null, null, config, tracer)
+  }
+}
+
+function createChownTags (config, tracer) {
+  return function chownTags (path, uid, gid) {
+    if (typeof path === 'number' || typeof uid !== 'number' || typeof gid !== 'number') {
+      return
+    }
+    const tags = makeFSTags(path, null, null, config, tracer)
+    tags['file.uid'] = uid.toString()
+    tags['file.gid'] = gid.toString()
+    return tags
+  }
+}
+
+function createFchownTags (config, tracer) {
+  return function fchownTags (fd, uid, gid) {
+    if (typeof this === 'object' && this !== null && this.fd) {
+      fd = this.fd
+    }
+    if (typeof fd !== 'number' || typeof uid !== 'number' || typeof gid !== 'number') {
+      return
+    }
+    const tags = makeFSTags(fd, null, null, config, tracer)
+    tags['file.uid'] = uid.toString()
+    tags['file.gid'] = gid.toString()
+    return tags
+  }
+}
+
+function getSymbolName (sym) {
+  return sym.description || sym.toString()
+}
+
+const tagMakers = {
+  open: createOpenTags,
+  close: createCloseTags,
+  readFile: createReadFileTags,
+  writeFile: createWriteFileTags,
+  appendFile: createAppendFileTags,
+  access: createPathTags,
+  copyFile: createCopyFileTags,
+  stat: createPathTags,
+  lstat: createPathTags,
+  fstat: createFDTags,
+  readdir: createPathTags,
+  opendir: createPathTags,
+  read: createFDTags,
+  write: createFDTags,
+  writev: createFDTags,
+  chmod: createChmodTags,
+  lchmod: createChmodTags,
+  fchmod: createFchmodTags,
+  chown: createChownTags,
+  lchown: createChownTags,
+  fchown: createFchownTags,
+  realpath: createPathTags,
+  readlink: createPathTags,
+  unlink: createPathTags,
+  symlink: createCopyFileTags,
+  link: createCopyFileTags,
+  rmdir: createPathTags,
+  rename: createCopyFileTags,
+  fsync: createFDTags,
+  fdatasync: createFDTags,
+  mkdir: createPathTags,
+  truncate: createPathTags,
+  ftruncate: createFDTags,
+  utimes: createPathTags,
+  futimes: createFDTags,
+  mkdtemp: createPathTags
+}
+
+function createWrapCb (tracer, config, name, tagMaker) {
+  const makeTags = tagMaker(config, tracer)
+  name = 'fs.' + name
+  return function wrapFunction (fn) {
+    return function wrappedFunction () {
+      const cb = arguments[arguments.length - 1]
+      if (typeof cb !== 'function') {
+        return fn.apply(this, arguments)
+      }
+      const tags = makeTags.apply(this, arguments)
+      if (!tags) {
```

Comment:
Returning falsy means tags can't be created (invalid input to the function) so we're passing it on through without tracing.

Reply:
What I mean is that it seems like `makeTags` could return falsy even if only some less important tags are missing. Just trying to make sure we don't stop tracing just because of a missing `file.*` tag for example.

Reply:
> it seems like makeTags could return falsy even if only some less important tags are missing

I tried to make that only happen when the arguments to the `fs` function in question are *invalid*, meaning the function will throw/reject/cb(err) anyway.

However.

I suppose those errors ought to be tracked anyway? Therefore we should continue tracing, just have missing tags?

Reply:
Definitely. Tracing should still occur if at least the core tags are available, especially in case of an error.